File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int daysNeed(vector<int>& weights,int load){
        int days = 1 , curr = 0 ;
        for(int i = 0 ; i < weights.size() ; i++){
            if(curr + weights[i] > load){
                days++;
                curr = weights[i];
            }
            else
                curr += weights[i];
        }
        return days;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int low = *max_element(weights.begin(),weights.end());
        int high = accumulate(weights.begin(), weights.end(), 0); 
        int ans = -1 ;
        while(low<=high){
            int mid = low + (high - low)/2;
            int day = daysNeed(weights,mid);
            if(day<=days){
                ans = mid;
                high = mid - 1 ; 
            }

            if(day>days)
                low = mid + 1;
            else
                high = mid - 1 ;
        }
        return ans ;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int low = 0 ;
        for(int w : weights){
            low = Math.max(low,w);
        }
        int high = 0 ;
        for(int weight : weights){
            high += weight ;
        }
        int ans = -1 ;
        while(low <= high){
            int mid = low + (high - low) / 2 ;
            int checkDays = 1 ;
            int sum = 0 ;
            for (int w : weights) {
                if (sum + w > mid) { // capacity exceeded
                checkDays++; // start new day
                sum = w;
                }
                else
                sum += w;
            }
        
            if(checkDays <= days){
                high = mid - 1 ;
            }
            else
                low = mid + 1 ;
        }
        return low ;
    }
}